### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by powlaa**
Links to our Code Reviews: 
https://github.com/Systems-Development-and-Frameworks/dojo/pull/5#pullrequestreview-537821742
https://github.com/Systems-Development-and-Frameworks/dojo/pull/5#pullrequestreview-537821051

### CODE REVIEW

**Review by powlaa — COMMENTED** [review]
I added apollo-client in the webapp and it now calls the graphql endpoints as well. Unfortunately storybook is currently not working because of this. It has issues with importing .gql files and we should probably mock the client there like we did in the tests right? 
With a little googling I found this: https://github.com/abhiaiyer91/apollo-storybook-decorator 
and this: https://github.com/storybookjs/storybook/issues/9581
but not sure if that will be helpful or not. 
Will look into this further but didn't manage to find a quick solution before the deadline.

File: backend/src/resolvers.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        },
+    },
+    Post: {
+        author: async (obj, args, context) => {
+            return await context.dataSources.db.getUser(obj.authorName);
+        },
+        votes: async (obj, args, context) => {
+            let values = Array.from(obj.voters.values());
+            let votes = values.reduce((sum, number) => sum + number, 0);
+            return votes;
```
Comment (jobegrabber) [inline backend/src/resolvers.js:62]: This is not going to be an issue for our miniature application, but calculating the number of votes on each request seems way to inefficient compared to how simple it is to calculate it in advance.

I don't know Apollo good enough to tell whether it can do some caching here, but even if it does I would try to circumvent caching if possible, because, you know
>There are only two hard things in Computer Science: cache invalidation and naming things.

File: backend/src/datasource.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    deletePost(id) {
+        const deletedPost = this.getPost(id);
+        this.posts = this.posts.filter((post) => post.id !== id);
+        return Promise.resolve(deletedPost);
+    }
+
+    upvotePost(id, user) {
+        return this.getPost(id).then((post) => {
+            post.voters.set(user, 1);
```
Comment (jobegrabber) [inline backend/src/datasource.js:56]: And yet another 
> :star: For preventing users to upvote the same post multiple times.

File: backend/src/datasource.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import { DataSource } from "apollo-datasource";
+import crypto from "crypto";
+
+export class Post {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString("hex");
```
Comment (jobegrabber) [inline backend/src/datasource.js:6]: When generating random IDs it's (in my experience) "business standard" to reuse a cryptographically sound library to generate UUIDs like [uuidjs/uuid](https://github.com/uuidjs/uuid)

File: backend/src/datasource.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        this.voters = new Map();
+        Object.assign(this, data);
```
Comment (jobegrabber) [inline backend/src/datasource.js:8]: I really, _really_ dislike this way of declaring constructors and initializing objects.
I can't really fault you for this because @roschaefer used and showed this off in the lecture, but this way you'll loose all the knowledge about which properties `Post` has and can be setup with..

File: backend/src/datasource.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    }
+}
+export class User {
+    constructor(name) {
+        this.name = name;
```
Comment (jobegrabber) [inline backend/src/datasource.js:13]: In comparsion to [here](https://github.com/Systems-Development-and-Frameworks/lichtow/pull/8/files#diff-10d67b04ad805bf8ca07ecb8e8488f88d1470956bcb26e612b6b7a940a3fd553R8), you can actually tell that `User` objects ought to have a `name`!

File: backend/src/datasource.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    constructor() {
+        super();
+        this.posts = [];
+        this.users = [];
+    }
+
+    initialize({ context }) {}
+
+    allUsers() {
+        return Promise.resolve(this.users);
```
Comment (jobegrabber) [inline backend/src/datasource.js:27]: Not sure why you implemented these methods as asynchronous.
Perhaps for forwards compatibility? I do expect actual DB connectors to have asynchronous APIs, though I don't know jack about Node.js..

File: backend/src/datasource.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    getUser(name) {
+        return Promise.resolve(this.users.find((user) => user.name === name));
```
Comment (jobegrabber) [inline backend/src/datasource.js:31]: Similar to calculating the number of votes on each requests, using `find` here (and in the other methods) has a worst-case complexity of O(n), though - as pretty much always in this assignment - there won't be any noticeable performance issue with this implementation.

Conceptually, `this.posts` and `this.users` are used to as lookup-tables and such could be also modelled as `Map` objects. This naturally prevents duplicates from being introduced, too.

File: backend/src/posts.spec.js [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        `;
+        const action = () =>
+            mutate({ mutation: WRITE_POST, variables: { post: { title: "Some post", author: { name: "Jonas" } } } });
+        const invalidUser = () =>
+            mutate({ mutation: WRITE_POST, variables: { post: { title: "Some post", author: { name: "INVALID" } } } });
+
+        it("throws error when user is invalid", async () => {
+            const {
+                errors: [error],
+            } = await invalidUser();
```
Comment (jobegrabber) [inline backend/src/posts.spec.js:76]: Nice way to destructure the response object!
I would add a test for `data` to be `null` though (in other places of `error` responses as well).

File: backend/src/posts.spec.js [hunk 9]
```diff
@@ -0,0 +1,278 @@
+import { gql } from "apollo-server";
+import { createTestClient } from "apollo-server-testing";
+import Server from "./server";
+import { InMemoryDataSource, User, Post } from "./datasource";
+
+let db;
+beforeEach(() => {
+    db = new InMemoryDataSource();
+    db.users.push(new User("Jonas"));
+});
```
Comment (jobegrabber) [inline backend/src/posts.spec.js:10]: Nice way to ensure atomic tests :+1:

File: backend/src/resolvers.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    Query: {
+        users: (parent, args, context) => context.dataSources.db.allUsers(),
+        posts: (parent, args, context) => context.dataSources.db.allPosts(),
+    },
+    Mutation: {
+        write: async (parent, args, context) => {
+            const newPost = {
+                title: args.post.title,
+                authorName: args.post.author.name,
+            };
```
Comment (jobegrabber) [inline backend/src/resolvers.js:13]: Personally, these resolvers contain too much logic to construct response objects, but that's a matter of taste I reckon.
Good code style in any case!

File: backend/src/resolvers.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            const author = await context.dataSources.db.getUser(newPost.authorName);
+            if (!author) {
+                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
+            }
+            return await context.dataSources.db.createPost(newPost);
+        },
```
Comment (jobegrabber) [inline backend/src/resolvers.js:19]: You could destructure `context` into `dataSources` right away and signal that you ignore `parent`:
```suggestion
        write: async (_, args, { dataSources }) => {
            const newPost = {
                title: args.post.title,
                authorName: args.post.author.name,
            };
            const author = await dataSources.db.getUser(newPost.authorName);
            if (!author) {
                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
            }
            return await dataSources.db.createPost(newPost);
        },
```

Comment (jobegrabber) [reply backend/src/resolvers.js:19]: Applies to other resolvers as well :slightly_smiling_face:

File: backend/src/resolvers.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            }
+            return await context.dataSources.db.downvotePost(id, name);
+        },
+    },
+    Post: {
+        author: async (obj, args, context) => {
+            return await context.dataSources.db.getUser(obj.authorName);
+        },
+        votes: async (obj, args, context) => {
+            let values = Array.from(obj.voters.values());
```
Comment (jobegrabber) [inline backend/src/resolvers.js:60]:
```suggestion
            let values = [...obj.voters.values()];
```
_should_ :tm: work as well

File: webapp/src/components/NewsList/NewsList.vue [hunk 13]
```diff
@@ -23,49 +25,144 @@
...
 import NewsItem from "../NewsItem/NewsItem.vue";
 import NewsForm from "../NewsForm/NewsForm.vue";
+import GET_POSTS from "../../graphql/Posts.gql";
+import WRITE_POST from "../../graphql/WritePost.gql";
+import DELETE_POST from "../../graphql/DeletePost.gql";
+import UPVOTE_POSTS from "../../graphql/UpvotePost.gql";
+import DOWNVOTE_POSTS from "../../graphql/DownvotePost.gql";
+
+//Use default user for now
+const DEFAULT_USER = "Jonas";
```
Comment (jobegrabber) [inline webapp/src/components/NewsList/NewsList.vue:35]: So much SCREAMING :fearful:

File: .github/workflows/backend.yml [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Uses Node.js ${{ matrix.node-version}}
+        uses: actions/setup-node@v1
+        with: 
+          node-version: ${{ matrix.node-version }}
+      - name: Install
+        run: npm install
+      - name: Run lint
+        run: npm run lint
+      - name: Run tests
+        run: npm run test
```
Comment (FlorianTh2) [inline .github/workflows/backend.yml:35]: Nice detailed github script

File: .github/workflows/backend.yml [hunk 15]
```diff
@@ -0,0 +1,35 @@
+
+name: CI Backend
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the origin/main branch
+on:
+  push:
+    branches: [ origin/main, exercise-3 ]
+  pull_request:
+    branches: [ origin/main, exercise-3 ]
```
Comment (FlorianTh2) [inline .github/workflows/backend.yml:10]: maybe setup for all branches, so you dont have to adjust it every new excercise

File: backend/src/datasource.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        });
+    }
+
+    downvotePost(id, user) {
+        return this.getPost(id).then((post) => {
+            post.voters.set(user, -1);
+            return post;
+        });
+    }
+}
```
Comment (FlorianTh2) [inline backend/src/datasource.js:67]: looks good to me

File: webapp/src/vue-apollo.js [hunk 17]
```diff
@@ -0,0 +1,78 @@
+import Vue from "vue";
```
Comment (jobegrabber) [inline webapp/src/vue-apollo.js:1]: Awesome you implemented the frontend-side as well! :clap: :smiley: 
Since I didn't dive into this on my own yet, I can't go review this in detail, though, sorry! :cry: 

However, I see one big issue: You seem to have copied this file in whole from this [stackoverflow answer](https://stackoverflow.com/a/61239384).
The copied code is definitely original work and since it seems to originate from the linked answer, as far as I can tell it's licensed under [CC-BY-SA](https://creativecommons.org/licenses/by-sa/3.0/) - which requires you to give reference to the original author and to license your derived work under the same (or compatible) license.
Both requirements are not met as far as i can see.

Apart from the legal issues (which don't really have practical implications to be honest), copying stuff without attribution is kind of questionable in academia :stuck_out_tongue:

File: backend/src/posts.spec.js [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+        it("returns post that is deleted", async () => {
+            await expect(deletePost()).resolves.toMatchObject({
+                errors: undefined,
+                data: {
+                    delete: { title: "Some post", id: postId, votes: 0, author: { name: "Jonas" } },
+                },
+            });
+        });
+    });
+});
```
Comment (FlorianTh2) [inline backend/src/posts.spec.js:278]: You have tested very carefully.

**Review by jobegrabber — CHANGES_REQUESTED** [review]
Don't think you added a test for recursively nested queries, but it should work..

And since you already reviewed my PR, you'll sure to get a
> :star: For requesting review and reviewing another teams PR according to the instructions.

Frontend things I couldn't review in-depth since I didn't forage into that area yet.
Apparently you copied a whole file from stackoverflow, which is not ideal (see the comment).

All in all great work though! Code is still easy to comprehend. :smiley: 
Only issues I see are based on individual taste :smile:

**Review by FlorianTh2 — APPROVED** [review]
i appreciate your work. It looks good to me.

File: webapp/src/vue-apollo.js (hunk 17 above)
Comment (powlaa) [reply webapp/src/vue-apollo.js:1]: I didn't copy the file from anywhere.  It was generated by calling 
`vue add apollo` which I used to quickly setup apollo-client.

File: backend/src/datasource.js (hunk 6 above)
Comment (powlaa) [reply backend/src/datasource.js:27]: Yeah it was just to see and figure out how it would work with asynchronous responses :)

File: webapp/src/vue-apollo.js (hunk 17 above)
Comment (jobegrabber) [reply webapp/src/vue-apollo.js:1]: Oh wow, I see! Then it's not an issue at all, of course :smile:

File: backend/src/datasource.js (hunk 4 above)
Comment (roschaefer) [reply backend/src/datasource.js:8]: That's a legitimate interest, if your build tools support it you could either use [instance fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields) or you could just do:
```js
const { some, white, listed, arguments } = data
Object.assign(this, { some, white, listed, arguments }); 
```

File: backend/src/resolvers.js (hunk 1 above)
Comment (roschaefer) [reply backend/src/resolvers.js:62]: You *could* setup a cache for this but then it's yet another cache (a piece of state) that you are responsible to keep up to date. My suggestion: Aim for absence of errors (e.g. avoid state) and solve performance problems when you have them. 

We developers have a tendency for premature optimization. Try to look for common performance problems (N+1, sequential execution, network), most often your application code is not the defining performance bottleneck. 

Also: Calculating the `votes` on the resolver has the performance advantage, that you only calculate it when requested:
```gql
{
  posts {
    title
    votes
  }
}
```
Will call `Post.votes` resolver, but:
```gql
{
  posts {
    title
  }
}
```
will not.

File: backend/src/server.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
+import { ApolloServer, gql } from "apollo-server";
+import typeDefs from "./typeDefs";
+import resolvers from "./resolvers";
+import { InMemoryDataSource, Post, User } from "./datasource";
+
+const db = new InMemoryDataSource();
+db.posts.push(new Post({ title: "Test Post", authorName: "Jonas" }));
+db.users.push(new User("Jonas"), new User("Paula"));
```
Comment (roschaefer) [inline backend/src/server.js:8]: I like that you seed your database here, in development environment only.

File: webapp/src/components/NewsList/NewsList.spec.js [hunk 20]
```diff
@@ -1,75 +1,119 @@
...
+    it("renders a Vue component", () => {
+        createComponent();
+        expect(wrapper.exists()).toBe(true);
+        expect(wrapper.vm.$apollo.queries.posts).toBeTruthy();
+    });
+
+    it("renders a loading block when query is in progress", () => {
+        createComponent();
+
+        expect(wrapper.find(".test-loading").exists()).toBe(true);
```
Comment (roschaefer) [inline webapp/src/components/NewsList/NewsList.spec.js:85]: :+1:

File: webapp/src/components/NewsList/NewsList.vue [hunk 21]
```diff
@@ -23,49 +25,144 @@
 <script>
 import NewsItem from "../NewsItem/NewsItem.vue";
 import NewsForm from "../NewsForm/NewsForm.vue";
+import GET_POSTS from "../../graphql/Posts.gql";
+import WRITE_POST from "../../graphql/WritePost.gql";
+import DELETE_POST from "../../graphql/DeletePost.gql";
+import UPVOTE_POSTS from "../../graphql/UpvotePost.gql";
+import DOWNVOTE_POSTS from "../../graphql/DownvotePost.gql";
```
Comment (roschaefer) [inline webapp/src/components/NewsList/NewsList.vue:32]: What about:
```suggestion
import { GET_POSTS } from "../../graphql/queries.gql";
import { WRITE_POST, DELETE_POST, UPVOTE_POSTS, DOWNVOTE_POSTS } from "../../graphql/mutations.gql";
```

File: webapp/src/components/NewsList/NewsList.vue [hunk 22]
```diff
@@ -23,49 +25,144 @@
...
       descending: true,
     };
   },
   methods: {
-    updateItem: function (item) {
-      if (item && item.id) {
-        this.items = this.items.filter((el) => el.id !== item.id);
-        this.items = [...this.items, item];
-      }
+    upvote: function (post) {
```
Comment (roschaefer) [inline webapp/src/components/NewsList/NewsList.vue:58]:
```suggestion
    upvote (post) {
```

File: webapp/src/components/NewsList/NewsList.vue [hunk 23]
```diff
@@ -23,49 +25,144 @@
...
+          upvote: {
+            __typename: "Post",
+            id: post.id,
+            title: post.title,
+            votes: post.votes + 1,
+            author: {
+              __typename: "User",
+              name: DEFAULT_USER,
+            },
+          },
```
Comment (roschaefer) [inline webapp/src/components/NewsList/NewsList.vue:81]: Good! I never used `optimisticResponse` myself yet. :wink:

File: webapp/src/components/NewsList/NewsList.vue [hunk 24]
```diff
@@ -23,49 +25,144 @@
...
+    downvote: function (post) {
+      this.$apollo.mutate({
+        mutation: DOWNVOTE_POSTS,
+        variables: {
+          id: post.id,
+          voter: { name: DEFAULT_USER },
+        },
+        update: (store, { data: { downvote } }) => {
+          const data = store.readQuery({ query: GET_POSTS });
+          let downvotedElement = data.posts.find((el) => el.id === downvote.id);
```
Comment (roschaefer) [inline webapp/src/components/NewsList/NewsList.vue:94]: Are you sure that this is even necessary? As far as I know, `vue-apollo` is smart enough to update existing types in the cache via the `ID`! That's why the `ID` should always be requested.

Comment (roschaefer) [reply webapp/src/components/NewsList/NewsList.vue:94]: Try to omit this and give me feedback if it has the desired effect or not.

File: webapp/src/vue-apollo.js [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+
+// Install the vue plugin
+Vue.use(VueApollo);
+
+const AUTH_TOKEN = "apollo-token";
+
+// Http endpoint
+const httpEndpoint = process.env.VUE_APP_GRAPHQL_HTTP || "http://localhost:4000/graphql";
+// Files URL root
+export const filesRoot = process.env.VUE_APP_FILES_ROOT || httpEndpoint.substr(0, httpEndpoint.indexOf("/graphql"));
```
Comment (roschaefer) [inline webapp/src/vue-apollo.js:13]:
```suggestion
export const filesRoot = process.env.VUE_APP_FILES_ROOT || httpEndpoint.replace('/graphql', '')
```

Comment (roschaefer) [reply webapp/src/vue-apollo.js:13]: What's the purpose of `filesRoot`? :thinking:

File: webapp/src/vue-apollo.js [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+                error.message
+            );
+        },
+    });
+
+    return apolloProvider;
+}
+
+// Manually call this when user log in
+export async function onLogin(apolloClient, token) {
```
Comment (roschaefer) [inline webapp/src/vue-apollo.js:53]: `@nuxtjs/apollo` has useful helpers for this: https://github.com/nuxt-community/apollo-module

**Review by roschaefer — APPROVED** [review]
excellent

Wow, this is the most sophisticated submission I've seen :muscle: Excellent!

This PR is textbook and I elect this PR as a reference solution for `#3`.

I've got only a few suggestions:
* Status checks on all branches as @FlorianTh2 pointed out.
* You don't need promises, but you already said you know and you want to use them for educational purposes.
* You could improve readability of your tests a little.


Here are your well deserved 10/10 :star:

:star: For correct folder structure, basic apollo server setup and it's linter in the CI/CD pipeline.

:star: For the `apollo-server-testing` setup and its integration in the CI/CD pipeline.

:star: For implementing a custom data source for apollo-server

For implementing resolvers for the [target schema](#target-schema) and the following features:
1. :star: For listing all posts
2. :star: For listing all users
3. :star: For creating a post
4. :star: For upvoting a post

:star: For preventing users to upvote the same post multiple times.

:star: For 'indefinitely' nestable queries.

:star: For requesting review and reviewing another teams PR according to the instructions.

Besides that you achieved all three :rocket:

:rocket: Implement and test resolvers for the mutations marked as `OPTIONAL`

:rocket: Setup and configure [apollo-client](https://apollo.vuejs.org/guide/installation.html#vue-cli-plugin) in your webapp

:rocket: Call the GraphQL endpoint from your webapp

File: webapp/src/components/NewsList/NewsList.vue (hunk 24 above)
Comment (powlaa) [reply webapp/src/components/NewsList/NewsList.vue:94]: You are right! It is indeed not necessary here. Only for write and delete it needs to be updated manually.

Comment (roschaefer) [reply webapp/src/components/NewsList/NewsList.vue:94]: `vue-apollo` is really smart, hugh? :wink: That's actually one of the most amazing benefits that comes with a typed schema, the client can cleverly auto-update it's cache and you don't have to do anything yourself.

Comment (roschaefer) [reply webapp/src/components/NewsList/NewsList.vue:94]: You can also install apollo client browser plugin (available for Firefox and Chromium) and observe the cache in real time.

Comment (powlaa) [reply webapp/src/components/NewsList/NewsList.vue:94]: That sounds awesome! Will definitely check it out! 😄
